File: demo-python-service/patterns/transaction/saga.py
from enum import Enum
from dataclasses import dataclass, field
from typing import Dict, List, Callable, Any
import uuid
import logging
from datetime import datetime

logger = logging.getLogger(__name__)

class SagaStatus(Enum):
    STARTED = "STARTED"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"
    COMPENSATING = "COMPENSATING"
    COMPENSATED = "COMPENSATED"

@dataclass
class SagaState:
    saga_id: str
    current_step: str
    status: SagaStatus
    context: Dict[str, Any] = field(default_factory=dict)
    completed_steps: List[str] = field(default_factory=list)
    created_at: datetime = field(default_factory=datetime.now)

class SagaOrchestrator:
    def __init__(self):
        self.sagas: Dict[str, SagaState] = {}
        self.steps: Dict[str, Callable] = {}
        self.compensations: Dict[str, Callable] = {}
    
    def register_step(self, step_name: str, action: Callable, compensation: Callable):
        self.steps[step_name] = action
        self.compensations[step_name] = compensation
    
    def start_saga(self, saga_type: str, context: Dict[str, Any]) -> str:
        saga_id = str(uuid.uuid4())
        saga = SagaState(
            saga_id=saga_id,
            current_step="START",
            status=SagaStatus.STARTED,
            context=context
        )
        self.sagas[saga_id] = saga
        logger.info(f"Started saga {saga_id} of type {saga_type}")
        return saga_id
    
    def execute_step(self, saga_id: str, step_name: str) -> bool:
        saga = self.sagas.get(saga_id)
        if not saga:
            return False
        
        try:
            if step_name in self.steps:
                result = self.steps[step_name](saga.context)
                saga.completed_steps.append(step_name)
                saga.current_step = step_name
                logger.info(f"Saga {saga_id} completed step {step_name}")
                return True
        except Exception as e:
            logger.error(f"Saga {saga_id} failed at step {step_name}: {e}")
            saga.status = SagaStatus.FAILED
            self.compensate(saga_id)
            return False
        
        return False
    
    def compensate(self, saga_id: str):
        saga = self.sagas.get(saga_id)
        if not saga:
            return
        
        saga.status = SagaStatus.COMPENSATING
        
        # Execute compensations in reverse order
        for step in reversed(saga.completed_steps):
            if step in self.compensations:
                try:
                    self.compensations[step](saga.context)
                    logger.info(f"Compensated step {step} for saga {saga_id}")
                except Exception as e:
                    logger.error(f"Compensation failed for step {step}: {e}")
        
        saga.status = SagaStatus.COMPENSATED
    
    def complete_saga(self, saga_id: str):
        saga = self.sagas.get(saga_id)
        if saga:
            saga.status = SagaStatus.COMPLETED
            logger.info(f"Saga {saga_id} completed successfully")

# Global saga orchestrator instance
saga_orchestrator = SagaOrchestrator()

# Example saga steps
def validate_order(context):
    order_id = context.get('order_id')
    if not order_id:
        raise ValueError("Order ID is required")
    context['validated'] = True
    return True

def compensate_validate_order(context):
    context['validated'] = False
    logger.info("Compensated order validation")

def reserve_inventory(context):
    item_id = context.get('item_id')
    quantity = context.get('quantity', 1)
    context['reserved_inventory'] = {'item_id': item_id, 'quantity': quantity}
    return True

def compensate_reserve_inventory(context):
    reserved = context.get('reserved_inventory')
    if reserved:
        logger.info(f"Released inventory: {reserved}")
    context['reserved_inventory'] = None

def process_payment(context):
    amount = context.get('amount')
    if not amount or amount <= 0:
        raise ValueError("Invalid payment amount")
    context['payment_processed'] = True
    return True

def compensate_process_payment(context):
    if context.get('payment_processed'):
        logger.info("Refunded payment")
    context['payment_processed'] = False

# Register saga steps
saga_orchestrator.register_step('validate_order', validate_order, compensate_validate_order)
saga_orchestrator.register_step('reserve_inventory', reserve_inventory, compensate_reserve_inventory)
saga_orchestrator.register_step('process_payment', process_payment, compensate_process_payment)

File: demo-node-service/patterns/transaction/saga.js
const EventEmitter = require('events');
const { v4: uuidv4 } = require('uuid');

const SagaStatus = {
    STARTED: 'STARTED',
    IN_PROGRESS: 'IN_PROGRESS',
    COMPLETED: 'COMPLETED',
    FAILED: 'FAILED',
    COMPENSATING: 'COMPENSATING',
    COMPENSATED: 'COMPENSATED'
};

class SagaStep {
    constructor(name, execute, compensate = null) {
        this.name = name;
        this.execute = execute;
        this.compensate = compensate;
    }
}

class SagaOrchestrator extends EventEmitter {
    constructor() {
        super();
        this.sagas = new Map();
        this.sagaDefinitions = new Map();
    }

    registerSaga(sagaType, steps) {
        this.sagaDefinitions.set(sagaType, steps);
    }

    async startSaga(sagaType, context = {}) {
        const sagaId = uuidv4();
        const steps = this.sagaDefinitions.get(sagaType);
        
        if (!steps) {
            throw new Error(`Saga type not found: ${sagaType}`);
        }

        const saga = {
            id: sagaId,
            type: sagaType,
            status: SagaStatus.STARTED,
            currentStep: 0,
            context: { ...context, sagaId },
            steps,
            completedSteps: [],
            createdAt: Date.now()
        };

        this.sagas.set(sagaId, saga);
        this.emit('saga-started', { sagaId, sagaType, context });

        // Start executing steps
        await this.executeNextStep(sagaId);
        return saga;
    }

    async executeNextStep(sagaId) {
        const saga = this.sagas.get(sagaId);
        if (!saga || saga.status !== SagaStatus.STARTED && saga.status !== SagaStatus.IN_PROGRESS) {
            return;
        }

        if (saga.currentStep >= saga.steps.length) {
            saga.status = SagaStatus.COMPLETED;
            this.emit('saga-completed', { sagaId: saga.id, context: saga.context });
            return;
        }

        saga.status = SagaStatus.IN_PROGRESS;
        const step = saga.steps[saga.currentStep];

        try {
            console.log(`Executing saga step: ${step.name} for saga: ${sagaId}`);
            const result = await step.execute(saga.context);
            
            // Update context with step result
            saga.context = { ...saga.context, ...result };
            saga.completedSteps.push(step.name);
            saga.currentStep++;

            this.emit('step-completed', { 
                sagaId, 
                stepName: step.name, 
                result, 
                context: saga.context 
            });

            // Execute next step
            await this.executeNextStep(sagaId);

        } catch (error) {
            console.error(`Saga step failed: ${step.name} for saga: ${sagaId}`, error);
            saga.status = SagaStatus.FAILED;
            saga.error = error.message;
            
            this.emit('step-failed', { 
                sagaId, 
                stepName: step.name, 
                error: error.message 
            });

            // Start compensation
            await this.compensate(sagaId);
        }
    }

    async compensate(sagaId) {
        const saga = this.sagas.get(sagaId);
        if (!saga) return;

        saga.status = SagaStatus.COMPENSATING;
        this.emit('saga-compensating', { sagaId });

        // Execute compensation in reverse order
        for (let i = saga.completedSteps.length - 1; i >= 0; i--) {
            const stepName = saga.completedSteps[i];
            const step = saga.steps.find(s => s.name === stepName);
            
            if (step && step.compensate) {
                try {
                    console.log(`Compensating step: ${stepName} for saga: ${sagaId}`);
                    await step.compensate(saga.context);
                    
                    this.emit('step-compensated', { 
                        sagaId, 
                        stepName, 
                        context: saga.context 
                    });
                } catch (error) {
                    console.error(`Compensation failed for step: ${stepName}`, error);
                    this.emit('compensation-failed', { 
                        sagaId, 
                        stepName, 
                        error: error.message 
                    });
                }
            }
        }

        saga.status = SagaStatus.COMPENSATED;
        this.emit('saga-compensated', { sagaId, context: saga.context });
    }

    getSagaStatus(sagaId) {
        return this.sagas.get(sagaId);
    }

    getAllSagas() {
        return Array.from(this.sagas.values());
    }

    getMetrics() {
        const sagas = Array.from(this.sagas.values());
        return {
            total: sagas.length,
            byStatus: sagas.reduce((acc, saga) => {
                acc[saga.status] = (acc[saga.status] || 0) + 1;
                return acc;
            }, {}),
            averageDuration: this.calculateAverageDuration(sagas)
        };
    }

    calculateAverageDuration(sagas) {
        const completedSagas = sagas.filter(s => 
            s.status === SagaStatus.COMPLETED || s.status === SagaStatus.COMPENSATED
        );
        
        if (completedSagas.length === 0) return 0;
        
        const totalDuration = completedSagas.reduce((sum, saga) => {
            return sum + (Date.now() - saga.createdAt);
        }, 0);
        
        return totalDuration / completedSagas.length;
    }
}

// Example saga steps
const createOrderProcessingSaga = () => [
    new SagaStep(
        'validate_order',
        async (context) => {
            const { orderId } = context;
            if (!orderId) throw new Error('Order ID is required');
            console.log(`Validating order: ${orderId}`);
            return { validated: true };
        },
        async (context) => {
            console.log(`Compensating order validation for: ${context.orderId}`);
        }
    ),
    new SagaStep(
        'reserve_inventory',
        async (context) => {
            const { itemId, quantity } = context;
            console.log(`Reserving inventory: ${itemId}, quantity: ${quantity}`);
            return { inventoryReserved: true, reservationId: uuidv4() };
        },
        async (context) => {
            console.log(`Releasing inventory reservation: ${context.reservationId}`);
        }
    ),
    new SagaStep(
        'process_payment',
        async (context) => {
            const { amount } = context;
            if (!amount || amount <= 0) throw new Error('Invalid payment amount');
            console.log(`Processing payment: $${amount}`);
            return { paymentProcessed: true, transactionId: uuidv4() };
        },
        async (context) => {
            console.log(`Refunding payment: ${context.transactionId}`);
        }
    )
];

module.exports = {
    SagaOrchestrator,
    SagaStep,
    SagaStatus,
    createOrderProcessingSaga
};